builtin/credential/cert/path_login.go

```diff
@@ -271,16 +272,27 @@ func (b *backend) verifyCredentials(ctx context.Context, req *logical.Request, d
 
 	// If trustedNonCAs is not empty it means that client had registered a non-CA cert
 	// with the backend.
+	var retErr error
 	if len(trustedNonCAs) != 0 {
 		for _, trustedNonCA := range trustedNonCAs {
 			tCert := trustedNonCA.Certificates[0]
 			// Check for client cert being explicitly listed in the config (and matching other constraints)
 			if tCert.SerialNumber.Cmp(clientCert.SerialNumber) == 0 &&
 				bytes.Equal(tCert.AuthorityKeyId, clientCert.AuthorityKeyId) {
 				matches, err := b.matchesConstraints(ctx, clientCert, trustedNonCA.Certificates, trustedNonCA, verifyConf)
-				if err != nil {
-					return nil, nil, err
+
+				// matchesConstraints returns an error when OCSP verification fails,
+				// but some other path might still give us success. Add to the
+				// retErr multierror, but avoid duplicates. This way, if we reach a
+				// failure later, we can give additional context.
+				//
+				// XXX: If matchesConstraints is updated to generate additional,
+				// immediately fatal errors, we likely need to extend it to return
+				// another boolean (fatality) or other detection scheme.
+				if err != nil && (retErr == nil || !errwrap.Contains(retErr, err.Error())) {
+					retErr = multierror.Append(retErr, err)
 				}
+
 				if matches {
```

Comment:
Could we add an additional protection `matches && err == nil`, just in case something returned a true, err which used to short-circuit failure in the past, we'd return the trustedNonCA now.

Reply:
Sure. While we're here, what's your thoughts on the second iteration? Here, we build up `matches`, but we only ever return the first element from it... Should we mirror this behavior, and avoid checking _all_ the chains, given the first (valid) match will succeed the request anyways...?

Reply:
Sure sounds okay to me.